chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleGridDataSet.java

```diff
@@ -0,0 +1,255 @@
+        super(name, grid.length + vals.length);
+        set(copy, grid, vals);
+    }
+
+    @Override
+    public double get(int dimIndex, int index) {
+        if (dimIndex < shape.length) {
+            return grid[dimIndex][values[0].getIndices(index)[dimIndex]];
+        }
+        return values[dimIndex - shape.length].getDouble(index);
+    }
+
+    @Override
+    public int getDataCount() {
+        return dataCount;
+    }
+
+    @Override
+    public int[] getShape() {
+        return shape;
+    }
+
+    @Override
+    public int getShape(final int dimIndex) {
+        return shape[dimIndex];
+    }
+
+    @Override
+    public double getGrid(int dimIndex, int index) {
+        if (dimIndex >= shape.length) {
+            throw new IndexOutOfBoundsException("Grid index out of bounds");
+        }
+        return grid[dimIndex][index];
+    }
+
+    @Override
+    public double[] getGridValues(int dimIndex) {
+        if (dimIndex >= shape.length) {
+            throw new IndexOutOfBoundsException("Grid index out of bounds");
+        }
+        return grid[dimIndex];
+    }
+
+    @Override
+    public double get(final int dimIndex, final int... indices) {
+        if (dimIndex < shape.length) {
+            return grid[dimIndex][indices[dimIndex]];
+        }
+        return values[dimIndex - shape.length].getDouble(indices);
+    }
+
+    @Override
+    public double[] getValues(int dimIndex) {
+        if (dimIndex >= shape.length) {
+            return values[dimIndex - shape.length].getElements();
+        }
+        return super.getValues(dimIndex); // return new list with full coordinates
+    }
+
+    @Override
+    public void set(final boolean copy, final double[][] grid, final double[]... vals) {
+        lock().writeLockGuard(() -> {
+            final int nDims = getDimension();
+            if (nDims != grid.length + vals.length) {
+                throw new IllegalArgumentException("grid + value dimensions must match dataset dimensions");
+            }
+            shape = IntStream.range(0, grid.length).map(i -> grid[i].length).toArray();
+            this.grid = copy ? new double[shape.length][] : grid;
+            dataCount = 1;
+            for (int i = 0; i < shape.length; i++) {
+                dataCount *= shape[i];
+                if (copy) {
+                    this.grid[i] = grid[i].clone();
+                }
+            }
+            values = new MultiArray[vals.length];
+            for (int i = shape.length; i < nDims; i++) {
+                if (vals[i - shape.length].length != dataCount) {
+                    throw new IllegalArgumentException("Dimension missmatch between grid and values");
+                }
+                values[i - shape.length] = new MultiArrayImpl<>(copy ? vals[i - shape.length].clone() : vals[i - shape.length], shape);
+            }
+        });
+        fireInvalidated(new UpdatedDataEvent(this));
+    }
+
+    @Override
+    public void set(GridDataSet another) {
```

Comment:
Codacy found an issue: [The method 'set(GridDataSet)' has an NPath complexity of 256, current threshold is 200](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5780327)